noir-projects/noir-protocol-circuits/crates/reset-kernel-lib/src/public_validation_request_processor.nr

```diff
@@ -0,0 +1,127 @@
+use crate::{
+    reset::{
+    non_existent_read_request::reset_non_existent_read_requests,
+    mutable_data_read_request::reset_mutable_data_read_requests, read_request::reset_read_requests
+},
+    nullifier_read_request_reset::NullifierReadRequestHints,
+    nullifier_non_existent_read_request_reset::NullifierNonExistentReadRequestHints,
+    public_data_read_request_reset::PublicDataReadRequestHints, types::public_data_hint::PublicDataHint
+};
+use dep::types::{
+    abis::{
+    kernel_circuit_public_inputs::PublicKernelCircuitPublicInputs,
+    public_data_update_request::PublicDataUpdateRequest, side_effect::SideEffectLinkedToNoteHash,
+    validation_requests::ValidationRequests
+},
+    constants::{MAX_NEW_NULLIFIERS_PER_TX, MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX},
+    hash::silo_nullifier, traits::is_empty,
+    utils::arrays::{array_merge, array_to_bounded_vec, assert_sorted_array}
+};
+
+struct PublicValidationRequestProcessor<N> {
+    validation_requests: ValidationRequests,
+    pending_nullifiers: [SideEffectLinkedToNoteHash; MAX_NEW_NULLIFIERS_PER_TX],
+    pending_public_data_writes: [PublicDataUpdateRequest; MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX],
+    nullifier_read_request_hints: NullifierReadRequestHints,
+    nullifier_non_existent_read_request_hints: NullifierNonExistentReadRequestHints,
+    nullifier_tree_root: Field,
+    public_data_read_request_hints: PublicDataReadRequestHints,
+    public_data_hints: [PublicDataHint; N],
+}
+
+impl<N> PublicValidationRequestProcessor<N> {
+    pub fn new(
+        public_inputs: PublicKernelCircuitPublicInputs,
+        nullifier_read_request_hints: NullifierReadRequestHints,
+        nullifier_non_existent_read_request_hints: NullifierNonExistentReadRequestHints,
+        nullifier_tree_root: Field,
+        public_data_read_request_hints: PublicDataReadRequestHints,
+        public_data_hints: [PublicDataHint; N]
+    ) -> Self {
+        let end_non_revertible = public_inputs.end_non_revertible;
+        let end = public_inputs.end;
+
+        let pending_nullifiers = array_merge(end_non_revertible.new_nullifiers, end.new_nullifiers);
+
+        let pending_public_data_writes = array_merge(
+            end_non_revertible.public_data_update_requests,
+            end.public_data_update_requests
+        );
+
+        PublicValidationRequestProcessor {
+            validation_requests: public_inputs.validation_requests,
+            pending_nullifiers,
+            pending_public_data_writes,
+            nullifier_read_request_hints,
+            nullifier_non_existent_read_request_hints,
+            nullifier_tree_root,
+            public_data_read_request_hints,
+            public_data_hints
+        }
+    }
+
+    pub fn validate(self) {
+        self.validate_nullifier_read_requests();
+        self.validate_nullifier_non_existent_read_requests();
+        self.validate_public_data_read_requests();
+    }
+
+    fn validate_nullifier_read_requests(self) {
```

Comment:
This and `validate_nullifier_non_existent_read_requests` below are copied from public tail.